### ISSUE TITLE
Auto revolve open-api docs from JsonResult type

### ISSUE THREAD

**Issue by mehdihadeli** [dotnet/aspnetcore#47630]
Hi,
Currently .net core 7 doesn't support auto revolve open-api docs from `JsonResult` type because it doesn't implement `IEndpointMetadataProvider`. Could you consider this in the next version?

**Comment by captainsafia** [comment]
The `JsonResult` type is an MVC-specific result type (as in: it's an implementation of `IActionResult`). Built-in `IEndpointMetadataProvider` implementations are applicable on the `IResult` types. The `IResult` analog for this type is the `JsonHttpResult` type which does contain an `IEndpointMetadataProvider` implementation. I'm inclined to not provide `IEndpointMetadataProvider` implementations for the `ActionResult` because it would be usless at the moment given https://github.com/dotnet/aspnetcore/issues/44988 the metadata wouldn't be processed anyways if you are in an MVC controller.

One thing to note here is that the `JsonResult` type is not generic so you'd either have to annotate it as an open-ended Dictionary type in Swagger or provide your own definitions if you know what the type is.

What kind of API are you using `JsonResult` in?

**Comment by mehdihadeli** [comment]
Thanks for your response. That was my mistake, I mean `JsonHttpResult<TValue>` and it implements `IResult` for using in minimal apis, but it doesn't implement `IEndpointMetadataProvider`. This maybe useful for someone like me. I use that in a generic endpoint like this, it will be great if support open-api with `JsonHttpResult<TValue>` response type automatically:

``` csharp
    public static RouteHandlerBuilder MapCommandEndpoint<TRequest, TResponse, TCommand, TCommandResult>(
        this IEndpointRouteBuilder builder,
        string pattern,
        int statusCode,
        Func<TRequest, TCommand>? mapRequestToCommand = null,
        Func<TCommandResult, TResponse>? mapCommandResultToResponse = null
    )
    where TRequest : class
    where TResponse : class
    where TCommandResult : class
    where TCommand : IRequest<TCommandResult>
    {
        return builder.MapPost(pattern, Handle);

        async Task<JsonHttpResult<TResponse>> Handle([AsParameters] HttpCommand<TRequest> requestParameters)
        {
            var (request, context, mediator, mapper, cancellationToken) = requestParameters;

            var command = mapRequestToCommand is not null ?
                              mapRequestToCommand(request) :
                              mapper.Map<TCommand>(request);
            var res = await mediator.Send(command, cancellationToken);

            var response = mapCommandResultToResponse is not null ?
                               mapCommandResultToResponse(res) :
                               mapper.Map<TResponse>(res);

            // https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/responses
            // https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/openapi?view=aspnetcore-7.0#multiple-response-types
            return TypedResults.Json(response, statusCode: statusCode);
        }
    }
```

**Comment by mitchdenny** [comment]
@mehdihadeli what were you hoping for in the current situation. You are returning a JsonHttpResult from the method with the status code being dynamically set at runtime. The Swagger document produced includes the endpoint but can't provide more detail about the status codes because that information isn't available at that point.

**Comment by mehdihadeli** [comment]
Just to generalize my response type, because in my endpoint there are multiple possibilities for the result, and putting all theme in the results is one approach, but I think `JsonHttpResult` also should work. When we want to write our response if `JsonHttpResult` has a `StatusCode` it could be translated to swagger metadata, but problem here is that `JsonHttpResult` doesn't implement `IEndpointMetadataProvider`

**Comment by mitchdenny** [comment]
Not quite. The `JsonHttpResult` instance that you are returning has a status code - but this is runtime information that is present when the delegate executes. It isn't really possible to reliably determine what the value of `statusCode` is inside the handler.

To conceptualize this a bit better imagine if your code instead set the value of `statusCode` by reading a file. It isn't possible for implementor of `IEndpointMetadataProvider` to find out what that value is unless it actually executes the code to read the file.

**Comment by mehdihadeli** [comment]
Thanks for your explanation, so this is not possible because `IEndpointMetadataProvider ` implemented a static `PopulateMetadata` and we don't have any control on the `JsonHttpResult` status code out there.

``` csharp
public static void PopulateMetadata(MethodInfo method, EndpointBuilder builder)
{
    builder.Metadata.Add(new ResponseMetadata(???, typeof(ProblemDetails)));
}
```
